File: 0966-vowel-spellchecker/0966-vowel-spellchecker.py
class Solution:
    def spellchecker(self, wordlist: List[str], queries: List[str]) -> List[str]:
        words = set(wordlist)
        capital = {}  # normalise case
        vowel = {}    # normalise vowels

        for word in wordlist:
            upper = word.upper()
            devowel = re.sub(r'[AEIOU]', '#', upper)
            capital.setdefault(upper, word)
            vowel.setdefault(devowel, word)

        for i in range(len(queries)):
            query = queries[i]
            if query in words: continue  # no change needed
            
            upper = query.upper()
            devowel = re.sub(r'[AEIOU]', '#', upper)
            if upper in capital:
                queries[i] = capital[upper]
            elif devowel in vowel:
                queries[i] = vowel[devowel]
            else:
                queries[i] = ""

        return queries

File: 0966-vowel-spellchecker/0966-vowel-spellchecker.java
class Solution {
    public String[] spellchecker(String[] wordlist, String[] queries) {
        var words = new HashSet<String>(Arrays.asList(wordlist));
        var capital = new HashMap<String, String>();    // normalise case
        var vowel = new HashMap<String, String>();      // normalise vowels

        for(var word : wordlist){
            String upper = word.toUpperCase(), devowel = upper.replaceAll("[AEIOU]", "#");
            capital.putIfAbsent(upper, word);
            vowel.putIfAbsent(devowel, word);
        }

        for(var i=0; i<queries.length; i++){
            var query = queries[i];
            if(words.contains(query)) continue; // no change needed
            String upper = query.toUpperCase(), devowel = upper.replaceAll("[AEIOU]", "#");
            if(capital.containsKey(upper))
                queries[i] = capital.get(upper);
            else if(vowel.containsKey(devowel))
                queries[i] = vowel.get(devowel);
            else
                queries[i] = "";
        }

        return queries;
    }
}